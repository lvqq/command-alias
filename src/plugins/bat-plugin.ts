import Plugin, { HandlerParams } from './plugin';

class BatPlugin extends Plugin {
  name: string;

  constructor() {
    super();
    this.name = 'bat';
  }

  handler(config: HandlerParams) {
    const aliaGroups = Object.keys(config).map(
      (groupName) =>
        `@REM ${groupName}\n${Object.keys(config[groupName])
          .map((key) => {
            const value = config[groupName][key];
            return `doskey ${key}=${value} $*`;
          })
          .join('\n')}`
    );
    return `@echo off\n\n${aliaGroups.join('\n\n')}\n`;
  }
}

export default BatPlugin;
