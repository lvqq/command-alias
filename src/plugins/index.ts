import ShPlugin from './sh-plugin';
import BatPlugin from './bat-plugin';
import Ps1Plugin from './ps1-plugin';

const plugins = [ShPlugin, BatPlugin, Ps1Plugin];

export default plugins;
