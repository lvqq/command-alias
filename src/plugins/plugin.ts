export type HandlerParams = Record<string, Record<string, string>>;

abstract class Plugin {
  abstract name: string;

  abstract handler(config: HandlerParams): string;
}

export default Plugin;
