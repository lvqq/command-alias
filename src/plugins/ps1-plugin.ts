import Plugin, { HandlerParams } from './plugin';

class Ps1Plugin extends Plugin {
  name: string;

  constructor() {
    super();
    this.name = 'ps1';
  }

  handler(config: HandlerParams) {
    const aliaGroups = Object.keys(config).map(
      (groupName) =>
        `# ${groupName}\n${Object.keys(config[groupName])
          .map((key) => {
            const value = config[groupName][key];
            return `function ${key} {\n  ${value} $args\n}`;
          })
          .join('\n\n')}`
    );
    return `#!/usr/bin/env pwsh\n\n${aliaGroups.join('\n\n')}\n`;
  }
}

export default Ps1Plugin;
