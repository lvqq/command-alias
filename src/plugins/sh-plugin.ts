import Plugin, { HandlerParams } from './plugin';

class ShPlugin extends Plugin {
  name: string;

  constructor() {
    super();
    this.name = 'sh';
  }

  handler(config: HandlerParams) {
    const aliaGroups = Object.keys(config).map(
      (groupName) =>
        `# ${groupName}\n${Object.keys(config[groupName])
          .map((key) => {
            const value = config[groupName][key];
            return `alias ${key}='${value}'`;
          })
          .join('\n')}`
    );
    return `${aliaGroups.join('\n\n')}\n`;
  }
}

export default ShPlugin;
