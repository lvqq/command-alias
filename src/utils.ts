export const JSONParse = (str: string | null) => {
  try {
    return str ? JSON.parse(str) : str;
  } catch (e) {
    return str;
  }
};
