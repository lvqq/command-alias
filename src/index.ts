import yaml from 'js-yaml';
import { existsSync, promises as fs } from 'node:fs';
import path from 'node:path';
import plugins from './plugins';
import { HandlerParams } from './plugins/plugin';
import { JSONParse } from './utils';

interface Options {
  filename?: string;
  plugins?: string[];
}

export const transform = async (
  source: string,
  outDir: string,
  options?: Options
) => {
  const config = (await fs.readFile(path.resolve(source))).toString();
  let alias: HandlerParams;
  if (source.endsWith('.yaml') || source.endsWith('yml')) {
    alias = yaml.load(config) as HandlerParams;
  } else {
    alias = JSONParse(config);
  }
  if (!existsSync(outDir)) {
    await fs.mkdir(outDir, { recursive: true });
  }
  await Promise.all(
    plugins.map(async (Plugin) => {
      const plugin = new Plugin();
      if (
        Array.isArray(options?.plugins) &&
        !options?.plugins?.includes(plugin.name)
      )
        return;
      const filename = path.resolve(
        outDir,
        `${options?.filename ?? 'alias'}.${plugin.name}`
      );
      await fs.writeFile(filename, plugin.handler(alias));
    })
  );
};
