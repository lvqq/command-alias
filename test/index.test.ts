/* eslint-disable no-restricted-syntax */
/* eslint-disable no-await-in-loop */
import { describe, expect, test } from 'vitest';
import { existsSync, promises as fs } from 'node:fs';
import { transform } from '../src';
import plugins from '../src/plugins';

describe('transform()', () => {
  test('generate alias from json', async () => {
    expect.assertions(plugins.length)
    await transform('./test/fixtures/alias.json', './test/output-json');
    for(const Plugin of plugins) {
      const { name } = new Plugin();
      const alias = (await fs.readFile(`./test/output-json/alias.${name}`)).toString();
      expect(alias).toMatchSnapshot();
    }
  });

  test('generate alias from yaml', async () => {
    expect.assertions(plugins.length)
    await transform('./test/fixtures/alias.yaml', './test/output-yaml');
    for(const Plugin of plugins) {
      const { name } = new Plugin();
      const alias = (await fs.readFile(`./test/output-yaml/alias.${name}`)).toString();
      expect(alias).toMatchSnapshot();
    }
  });

  test('generate alias with specified plugin', async () => {
    expect.assertions(plugins.length)
    await transform('./test/fixtures/alias.yaml', './test/output-specified', { plugins: ['sh'] });
    for(const Plugin of plugins) {
      const { name } = new Plugin();
      if (name === 'sh') {
        const alias = (await fs.readFile(`./test/output-specified/alias.${name}`)).toString();
        expect(alias).toMatchSnapshot();
      } else {
        expect(existsSync(`./test/output-specified/alias.${name}`)).toBeFalsy();
      }
    }
  });
});
