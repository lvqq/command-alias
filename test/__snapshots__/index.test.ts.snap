// Vitest Snapshot v1

exports[`generateCommandAliasByPlugin() > generate alias from json 1`] = `
"# part1
alias sa='short alias'

# part2
alias saa='short alias with optionA'
alias sab='short alias with optionB'
"
`;

exports[`generateCommandAliasByPlugin() > generate alias from json 2`] = `
"@echo off

@REM part1
doskey sa=short alias $*

@REM part2
doskey saa=short alias with optionA $*
doskey sab=short alias with optionB $*
"
`;

exports[`generateCommandAliasByPlugin() > generate alias from json 3`] = `
"#!/usr/bin/env pwsh

# part1
function sa {
  short alias $args
}

# part2
function saa {
  short alias with optionA $args
}

function sab {
  short alias with optionB $args
}
"
`;

exports[`generateCommandAliasByPlugin() > generate alias from yaml 1`] = `
"# part1
alias sa='short alias'

# part2
alias saa='short alias with optionA'
alias sab='short alias with optionB'
"
`;

exports[`generateCommandAliasByPlugin() > generate alias from yaml 2`] = `
"@echo off

@REM part1
doskey sa=short alias $*

@REM part2
doskey saa=short alias with optionA $*
doskey sab=short alias with optionB $*
"
`;

exports[`generateCommandAliasByPlugin() > generate alias from yaml 3`] = `
"#!/usr/bin/env pwsh

# part1
function sa {
  short alias $args
}

# part2
function saa {
  short alias with optionA $args
}

function sab {
  short alias with optionB $args
}
"
`;

exports[`generateCommandAliasByPlugin() > generate alias with specified plugin 1`] = `
"# part1
alias sa='short alias'

# part2
alias saa='short alias with optionA'
alias sab='short alias with optionB'
"
`;
